[65138] Do not seed if required references are missing

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/65138

# What are you trying to accomplish?

Avoid error when seeding on an instance having deleted all demo projects and all statuses (additional statuses were created) and having no workflows (because statuses were deleted, workflows were deleted too by cascade).

# What approach did you choose and why?

The seeders now have the ability to declare the attributes containing references to other objects. If those references are missing, the seeder will not run.

This is used on some seeders to avoid raising an error when seeding on an instance without any workflows (see bug 65138). The seeding is incomplete, but at least it does not raise during the seeding phase.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
[65138] Do not seed if required references are missing

https://community.openproject.org/wp/65138

The seeders now have the ability to declare the attributes containing
references to other objects. If those references are missing, the seeder
will not run.

This is used on some seeders to avoid raising an error when seeding on
an instance without any workflows (see bug 65138)

oliverguenther reviewed (approved):
I couldn't reproduce the linked error anymore 👍 I wonder if the seeder should output somewhere that it didn't run _because of_ the missing references? I would argue in most cases on an initial seed, that would be an error and we want to minimize the risk that it silently skips?

cbliard commented:
> I wonder if the seeder should output somewhere that it didn't run _because of_ the missing references?

I thought about it. That can be added.

> I would argue in most cases on an initial seed, that would be an error and we want to minimize the risk that it silently skips?

That's all the issue with seeding currently: it does not know it's an initial seed, it only guesses.

An alternate approach could be to never fail, so even if we have one of these pesky `ArgumentError` then it would still continue the upgrade. And we then add a parameter to ask for failure if something goes wrong, and we would set this parameter to 1 on tenant creation or on installation.

I can add this as part of this bug, or in another PR. What would be best @oliverguenther ?

cbliard committed:
Fix flaky spec

spec: `./modules/team_planner/spec/features/team_planner_add_existing_work_packages_spec.rb:176`
failing job: https://github.com/opf/openproject/actions/runs/15991857052/job/45107671723

Moving by 800 pixels was too imprecise: it was moved to Wednesday
instead of being moved to Tuesday as intended. It was ok when the
element was not fully loaded, but not when it was. Probably its
dimensions changed a bit and that was enough to fall on the wrong day...

The fix is to drag from the card to the target date column header, with
an y offset of the header's height. This way it's dragged exactly on the
day area below the day header. It works even if we change the default
dimensions of everything.
